### PR TITLE
Expose miniz to Python

### PR DESCRIPTION
**Pull request by driazati**
Stacked PRs
 * #29232 - Add zipfile serialization
 * #29244 - Use custom CRC
 * **#29228 - Expose miniz to Python**

This adds the miniz wrapper to Python along with some functionality so that it can operate on both files and buffers. Python's `zipfile` module is pretty slow (see #26573), but miniz solves most of the perf issues.

Differential Revision: [D18330945](https://our.internmc.facebook.com/intern/diff/18330945/)

### CODE REVIEW

File: torch/csrc/jit/init.cpp [hunk 1]
```diff
@@ -406,16 +406,41 @@ void initJITBindings(PyObject* module) {
...
+          "write_record",
+          [](PyTorchStreamWriter& self,
+             const std::string& name,
+             uintptr_t data,
+             size_t size) {
+            return self.writeRecord(
+                name, reinterpret_cast<const char*>(data), size);
+          });
+
+  m.def("make_bytes", [](py::capsule data, size_t len) {
```
Comment (zdevito) [inline torch/csrc/jit/init.cpp]: My overall impression here is that this API is hard to follow. There are a lot of crosses between python and C++ (make_bytes, read_into)  that have to be tracked down. Our experience with the torchscript/python bindings is that this stuff grows and becomes inscrutable. The interface between Python and C++ should be as narrow as possible.

File: caffe2/serialize/inline_container.h [hunk 2]
```diff
@@ -100,6 +101,7 @@ class CAFFE2_API PyTorchStreamReader final {
  public:
   explicit PyTorchStreamReader(const std::string& file_name);
   explicit PyTorchStreamReader(std::istream* in);
+  explicit PyTorchStreamReader(ReaderFunc in, SeekerFunc seeker, size_t size);
```
Comment (zdevito) [inline caffe2/serialize/inline_container.h]: This feels redundant withe ReadAdapterInterface. Both abstract the ability to read. I'd prefer that we only had ReadAdapterInterface. (In fact, I'd prefer that PyTorchStreamReader had only one constructor and the differences between taking a file, a stream, or something else were abstracted in ReadAdapterInterface.

File: torch/serialization.py [hunk 3]
```diff
@@ -207,6 +207,46 @@ def __exit__(self, *args):
             self.file_like.close()
 
 
+class _open_zipfile_writer(_open_file_like):
```
Comment (zdevito) [inline torch/serialization.py]: There are 4 classes `{buffer, file} x {input, output}` hidden in these two class, making the implementation have excessive if statements. It would make more sense to make this _open_file_like interface instanciate the right class on construction rather than defer it to later actions.

**Review by zdevito — COMMENTED** [review]
It makes sense to expose a faster zip implementation, but I feel like this could be abstracted better.

File: caffe2/serialize/inline_container.h (hunk 2 above)
Comment (driazati) [reply caffe2/serialize/inline_container.h]: `PyTorchStreamReader` just forwards these to construct a `ReadAdapterInterface` (`FuncAdapter`), are you saying this should happen earlier? Doing it as is also allows `PyTorchStreamReader` to be bound directly to Python without any custom weirdness around its constructor

File: torch/csrc/jit/init.cpp (hunk 1 above)
Comment (driazati) [reply torch/csrc/jit/init.cpp]: I agree but I don't think it's avoidable here within miniz's interface since Python doesn't have built in utilities to work with pre-allocated buffers. The pathway of Python  (torch.save/load) -> C++ (miniz) -> Python (for a generic `pickle`) -> C++ (for writing to a buffer) was the only way I found that worked

File: torch/serialization.py (hunk 3 above)
Comment (driazati) [reply torch/serialization.py]: Completely separating these apart will result in something like 8 classes (4 for zip files, 2 for regular files, plus probably some base classes) which seems way more complex than the current implementation.

File: caffe2/serialize/inline_container.h (hunk 2 above)
Comment (zdevito) [reply caffe2/serialize/inline_container.h]: PyTorchStreamReader should not need an additional constructor. The implementation of a ReadAdapterInterface that would work with python objects can be put directly in the binding for the constructor of PythonStreamReader.

File: torch/csrc/jit/init.cpp (hunk 1 above)
Comment (zdevito) [reply torch/csrc/jit/init.cpp]: I would expect to see an API for writing a zip file, and calls from Python to use that API. It is fine if the bound methods for that API take python objects and there is some code in C++ to convert the python object to a buffer), but directly exposing things like make_bytes and swapping back and forth does not make it easy to follow.

File: torch/serialization.py (hunk 3 above)
Comment (zdevito) [reply torch/serialization.py]: I don't think this needs to be 8 separate classes, but the current way it is organized is confusing because every function starts with an if statement. It would be a good idea to ask what the responsibility of a _open_file_like object actually is, and build object oriented abstractions for it.

**Review by zdevito — COMMENTED** [review]
Comments inline

File: caffe2/serialize/func_adapter.h [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+namespace serialize {
+
+// Read a number of bytes into a buffer and return the number of bytes actually
+// read
+using ReaderFunc = std::function<size_t(void*, size_t)>;
+
+// Used to seek the buffer to the specified position and returns the position
+using SeekerFunc = std::function<size_t(size_t)>;
+
+class CAFFE2_API FuncAdapter final : public ReadAdapterInterface {
```
Comment (ezyang) [inline caffe2/serialize/func_adapter.h]: What is a FuncAdapter

File: caffe2/serialize/func_adapter.h [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+#include <functional>
+#include "c10/macros/Macros.h"
+#include "caffe2/serialize/read_adapter_interface.h"
+
+namespace caffe2 {
+namespace serialize {
+
+// Read a number of bytes into a buffer and return the number of bytes actually
+// read
+using ReaderFunc = std::function<size_t(void*, size_t)>;
```
Comment (ezyang) [inline caffe2/serialize/func_adapter.h]: `std::function` is notoriously slow. Do we know that it's use here is acceptable?

File: torch/csrc/jit/init.cpp [hunk 6]
```diff
@@ -406,23 +407,65 @@ void initJITBindings(PyObject* module) {
...
-          "write_record",
-          [](PyTorchStreamWriter& self,
-             const std::string& name,
-             const char* data,
-             size_t size) { return self.writeRecord(name, data, size); })
-      .def("write_end_of_file", &PyTorchStreamWriter::writeEndOfFile);
+      .def(py::init([](const py::object &buffer) {
+        auto writer_func = [=](const void *data, size_t size) {
+          auto bytes = py::bytes(reinterpret_cast<const char *>(data), size);
+          buffer.attr("write")(std::move(bytes));
```
Comment (ezyang) [inline torch/csrc/jit/init.cpp]: Phew, capture by value!

File: torch/csrc/jit/init.cpp [hunk 7]
```diff
@@ -406,23 +407,65 @@ void initJITBindings(PyObject* module) {
...
+              uintptr_t data, size_t size) {
+             return self.writeRecord(name, reinterpret_cast<const char *>(data),
+                                     size);
+           });
 
   py::class_<PyTorchStreamReader>(m, "PyTorchFileReader")
       .def(py::init<std::string>())
-      .def("get_record", [](PyTorchStreamReader& self, const std::string& key) {
+      .def(py::init([](const py::object &buffer) {
+        // Get size
```
Comment (ezyang) [inline torch/csrc/jit/init.cpp]: Is it required to know size for the reader protocol? This means you can't use this reader for true streams, where seeking to the end would result in materializing all the data.

Comment (ezyang) [reply torch/csrc/jit/init.cpp]: (I guess the protocol here requires random access seeking, so this is fine)

**Review by ezyang — APPROVED** [review]
Looks reasonable, but the PR description doesn't really reflect the contents of the PR.

File: torch/csrc/jit/init.cpp [hunk 8]
```diff
@@ -406,23 +407,65 @@ void initJITBindings(PyObject* module) {
...
+        auto current = buffer.attr("tell")();
+        buffer.attr("seek")(current, py::module::import("os").attr("SEEK_END"));
+        size_t size = py::cast<size_t>(buffer.attr("tell")());
+        buffer.attr("seek")(current);
+
+        // Make function to read bytes from the buffer
+        auto reader_func = [=](void* output_buffer, size_t size) {
```
Comment (zdevito) [inline torch/csrc/jit/init.cpp]: The FuncAdaptor abstraction seems redundant. ReadAdapterInterface can be subclassed right here to implement read/size directly.  Otherwise, it gets confusing to maintain since there are two adapters for the stream writer.

**Review by zdevito — APPROVED** [review]
This looks good now, but can we please use the ReadAdapterInterface directly?

File: torch/csrc/jit/init.cpp (hunk 7 above)
Comment (driazati) [reply torch/csrc/jit/init.cpp]: Miniz requires knowing the size of whatever its data buffer is (I guess so it can jump to the central directory of the zip file
